File: Examples/Bayesian/BayesExample2.m
% Example of Bayesian estimation by MCMC
% sampling from exponential(theta) distribution
% lognormal prior

% Shows how MCMC can be used to get posterior mean

% explore different sample sizes, different true thetas


function BayesExample2
	close all;
	n = 30;    % sample size. NOTE: likelihood is computed directly, below. When sample size grows, 
                % IT DISAPPEARS! So, keep <= 2000
	truetheta = 3; % true theta
	y = exprnd(ones(n,1)*truetheta); % sample from exponential(theta)

	S = 5000;
	theta = 2.8; % start value for theta
	tuning = 0.7; % tunes the acceptance rate. Lowering increases acceptance 
                % try to get acceptance rate to be about 0.25 or so
	thetas = zeros(S,1);
	accepts = zeros(S,1);
	for s = 1:S
		thetastar =  proposal(theta, tuning);
		num = likelihood(y, thetastar)*prior(thetastar)*proposal_density(theta, thetastar, tuning);
		den = likelihood(y, theta)*prior(theta)*proposal_density(thetastar, theta, tuning);
		crit = num / den;
		accept = crit > rand(1,1);
		if accept
			theta = thetastar;
		end
		thetas(s,:) = theta;
		accepts(s,:) = accept;
	end
	plot(thetas);
	hold on;
	pm = mean(thetas(500:S,:));
	ps = std(thetas(500:S,:));
	plot([0; S], [pm; pm], 'g');
	plot([0; S], [truetheta; truetheta], 'r');
	legend('chain', 'posterior mean', 'true theta');
	fprintf('posterior mean %f\n', pm);
	fprintf('posterior std. dev.  %f\n', ps);
	fprintf('acceptance rate %f\n', mean(accepts));
	%print -dpng BayesExample2.png;
end


% the prior is lognormal
function p = prior(theta)
	p = lognpdf(theta, 1, 1);
end

% the likelihood function
function dens = likelihood(y, theta)
	% scale is used to try keep the likelihood from going to zero
    % as the sample size increases. Professional software would do
    % this better.
    % it cancels out of the acceptance formula
    scale = 1/0.165;
    dens = scale*(1./theta).*exp(-y./theta);
    dens = prod(dens); % independent obsn
end

% the proposal density: random walk lognormal
function f = proposal_density(thetastar, theta, tuning)
	f = lognpdf(thetastar, log(theta), tuning);
end
% the proposal: random walk lognormal
function thetastar = proposal(theta, tuning)
	thetastar = lognrnd(log(theta), tuning);
end

File: Examples/Bayesian/BayesExample2.jl
# Example of Bayesian estimation by MCMC
# sampling from exponential(theta) distribution
# lognormal prior

# play with n, and observe how posterior quantiles narrow
# IMPPORTANT: you need to reduce tuning when n increases,
# try to keep the acceptance rate around 40% or so.

using Distributions, Econometrics, Plots
pyplot()

function main()
	n = 30    # sample size
	truetheta = 3 # true theta
    y = rand(Exponential(truetheta), n) # sample from exponential(theta)

	chainlength = 100000
    burnin = 1000
	theta = 2.8     # start value for theta
	tuning = 0.25    # tunes the acceptance rate. Lowering increases acceptance 
                    # try to get acceptance rate to be about 0.25 or so
    chain = makechain(theta, chainlength, burnin, tuning, y)
	thetas = chain[:,1]    
    plot(thetas[end-1000:end],show=true)
    #savefig("chain.svg")
    x, d = npdensity(thetas)
    plot(x,d, show=true)
    #savefig("posterior.svg")
	dstats(chain)
end

# the prior is lognormal(1,1)
function prior(theta)
    d = LogNormal(1.0,1.0)
    p = pdf.(d, theta)
end

# the log likelihood function
function likelihood(y, theta)
    d = Exponential(theta)
    logdens = sum(logpdf.(d, y))
end

# the proposal density: random walk lognormal)
function prop_dens(trial, theta, tuning)
    d = LogNormal(log(theta), tuning)
	den = pdf(d, trial)
end

# the proposal: random walk lognormal
function proposal(theta, tuning)
    d = LogNormal(log(theta), tuning)
	trial = rand(d)
end

function makechain(theta, reps, burnin, tuning, data)
    Lcurrent = likelihood(data, theta)
    chain = zeros(reps, size(theta,1) + 1)
    for rep = 1:reps+burnin
        trial = proposal(theta, tuning)
        Ltrial = likelihood(data, trial)
        accept = 0
        # MH accept/reject (simple form for symmetric proposal)
        if rand() < (
                    exp(Ltrial-Lcurrent) * 
                    prior(trial)/prior(theta)*
                    prop_dens(theta, trial, tuning) / prop_dens(trial, theta, tuning)
                    )
            theta = trial
            Lcurrent = Ltrial
            accept = 1
        end
        if rep > burnin
            chain[rep-burnin,:] = [theta accept]
        end    
    end
    return chain
end

main()
